modules/setting/config_provider.go

```diff
@@ -4,20 +4,150 @@
+	Name() string
+	MapTo(interface{}) error
+	HasKey(key string) bool
+	NewKey(name, value string) (*ini.Key, error)
+	Key(key string) *ini.Key
+	Keys() []*ini.Key
+	ChildSections() []*ini.Section
+}
+
 // ConfigProvider represents a config provider
 type ConfigProvider interface {
-	Section(section string) *ini.Section
-	NewSection(name string) (*ini.Section, error)
-	GetSection(name string) (*ini.Section, error)
+	Section(section string) ConfigSection
+	NewSection(name string) (ConfigSection, error)
+	GetSection(name string) (ConfigSection, error)
+	DeleteSection(name string) error
+	Save() error
+}
+
+type iniFileConfigProvider struct {
+	*ini.File
+	p          string // the ini file path
+	newFile    bool   // whether the file exists
+	allowEmpty bool   // whether allow empty
+}
+
+// NewEmptyConfigProvider create a new empty config provider
+func NewEmptyConfigProvider() ConfigProvider {
+	return &iniFileConfigProvider{
+		File:    ini.Empty(),
+		newFile: true,
+	}
+}
+
+func newConfigProviderFromData(bs []byte) (ConfigProvider, error) {
+	cfg, err := ini.Load(bs)
+	if err != nil {
+		return nil, err
+	}
+	return &iniFileConfigProvider{
+		File:    cfg,
+		newFile: true,
+	}, nil
+}
+
+// newConfigProviderFromFile load configuration from file.
+// NOTE: do not print any log except error.
+func newConfigProviderFromFile(customConf string, allowEmpty bool, extraConfig string) (*iniFileConfigProvider, error) {
+	cfg := ini.Empty()
+	newFile := true
+
+	if customConf != "" {
+		isFile, err := util.IsFile(customConf)
+		if err != nil {
+			return nil, fmt.Errorf("unable to check if %s is a file. Error: %v", customConf, err)
+		}
+		if isFile {
+			if err := cfg.Append(customConf); err != nil {
+				return nil, fmt.Errorf("failed to load custom conf '%s': %v", customConf, err)
+			}
+			newFile = false
+		}
+	}
+
+	if newFile && !allowEmpty {
+		return nil, fmt.Errorf("unable to find configuration file: %q, please ensure you are running in the correct environment or set the correct configuration file with -c", CustomConf)
+	}
+
+	if extraConfig != "" {
+		if err := cfg.Append([]byte(extraConfig)); err != nil {
+			return nil, fmt.Errorf("unable to append more config: %v", err)
+		}
+	}
+
+	cfg.NameMapper = ini.SnackCase
+	return &iniFileConfigProvider{
+		File:       cfg,
+		p:          customConf,
+		newFile:    newFile,
+		allowEmpty: allowEmpty,
+	}, nil
+}
+
+func (p *iniFileConfigProvider) Section(section string) ConfigSection {
+	return p.File.Section(section)
+}
+
+func (p *iniFileConfigProvider) NewSection(name string) (ConfigSection, error) {
+	return p.File.NewSection(name)
+}
+
+func (p *iniFileConfigProvider) GetSection(name string) (ConfigSection, error) {
+	return p.File.GetSection(name)
+}
+
+func (p *iniFileConfigProvider) DeleteSection(name string) error {
+	p.File.DeleteSection(name)
+	return nil
+}
+
+// Save save the content into file
+func (p *iniFileConfigProvider) Save() error {
+	if p.p == "" {
```

Comment:
It seems difficult to understand what is `p.p` when reading the code.

Reply:
done.